File: frontend/src/app/services/favorite.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Globals } from '../global/globals';
import { ErrorHandlerService } from './error-handler.service';
import { DeckSimple } from '../dtos/deckSimple';
import { map, catchError } from 'rxjs/operators';
import { Observable } from 'rxjs';
import { Page } from '../dtos/page';
import { Pageable } from '../dtos/pageable';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root',
})
export class FavoriteService {
  constructor(
    private httpClient: HttpClient,
    private globals: Globals,
    private errorHandler: ErrorHandlerService,
    private authService: AuthService
  ) {}

  /**
   * Add a deck to the favorites of the logged in user
   *
   * @param deckId deck which should be added as favorite
   */
  addFavorite(deckId: number): Observable<DeckSimple> {
    return this.httpClient
      .put<DeckSimple>(this.getFavoriteUri(deckId), {})
      .pipe(
        catchError(this.errorHandler.handleError('Could not add favorite'))
      );
  }

  /**
   * Get page of favorites of the logged in user
   */
  getFavorites(pageable: Pageable): Observable<Page<DeckSimple>> {
    const params = pageable.toHttpParams();

    return this.httpClient
      .get<Page<DeckSimple>>(this.getFavoriteUri(), { params })
      .pipe(
        catchError(this.errorHandler.handleError('Could not get favorites'))
      );
  }

  /**
   * Check if a deck is in the favorites of the logged in user
   *
   * @param deckId deck which should be checked
   */
  hasFavorite(deckId: number): Observable<boolean> {
    // return false if the status is 404, else true for a successful response
    return this.httpClient.get<void>(this.getFavoriteUri(deckId)).pipe(
      catchError(this.errorHandler.catchStatus(404, false)),
      catchError(
        this.errorHandler.handleError('Could not check if deck is a favorite')
      ),
      map((val) => val !== false)
    );
  }

  /**
   * Remove deck from the favorites of the logged in user
   */
  removeFavorite(deckId: number): Observable<void> {
    return this.httpClient
      .delete<void>(this.getFavoriteUri(deckId))
      .pipe(
        catchError(this.errorHandler.handleError('Could not add favorite'))
      );
  }

  private getFavoriteUri(deckId?: number): string {
    const userId = this.authService.getUserId();
    if (typeof userId !== 'number') {
      console.error('Invalid userId', userId);
      throw new Error('Invalid userId for getFavoriteUri');
    }

    return (
      `${this.globals.backendUri}/users/${userId}/favorites` +
      (deckId ? `/${deckId}` : '')
    );
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/FavoriteService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Deck;
import at.ac.tuwien.sepm.groupphase.backend.exception.AuthenticationRequiredException;
import at.ac.tuwien.sepm.groupphase.backend.exception.DeckNotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.InsufficientAuthorizationException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ConflictException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;


public interface FavoriteService {
    /**
     * Add a deck as favorite of the user
     *
     * @param userId id of the user to which it should be added
     * @param deckId id of the favorite deck
     * @return created favorite
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the authenticated user does not match the {@code userId}
     * @throws DeckNotFoundException if no deck with this id exists
     * @throws ConflictException when the deck is already a favorite of the user
     */
    Deck addFavorite(Long userId, Long deckId);

    /**
     * Get favorites of the user
     *
     * @param userId id of the user to get the favorites from
     * @param pageable paging parameters
     * @return page containing favorites
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the authenticated user does not match the {@code userId}
     */
    Page<Deck> getFavorites(Long userId, Pageable pageable);

    /**
     * Check if a deck is already a favorite of the user
     *
     * @param userId id of the user to get the favorites from
     * @param deckId id of the deck which is possibly a favorite
     * @return true iff deck is a favorite of the user
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the authenticated user does not match the {@code userId}
     * @throws DeckNotFoundException if no favorite with this {@code deckId} exists
     */
    boolean hasFavorite(Long userId, Long deckId);

    /**
     * Remove a deck from the favorites of the user
     *
     * @param userId id of the user from which it should be removed
     * @param deckId id of the deck which should be removed
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the authenticated user does not match the {@code userId}
     * @throws DeckNotFoundException if no favorite with this {@code deckId} exists
     */
    void removeFavorite(Long userId, Long deckId);
}
